src/foam/nanos/medusa/CompactionDAO.js

```diff
@@ -0,0 +1,413 @@
+
+        // TODO: start in own thread
+        execute(x);
+        return obj;
+      }
+      return getDelegate().cmd_(x, obj);
+      `
+    },
+    {
+      name: 'maybeBlock',
+      args: 'X x',
+      javaCode: `
+      return getBlocking();
+      `
+    },
+    {
+      name: 'execute',
+      args: 'X x',
+      javaCode: `
+      Logger logger = Loggers.logger(x, this);
+      logger.info("start");
+      ClusterConfigSupport support = (ClusterConfigSupport) x.get("clusterConfigSupport");
+
+      try {
+        // block
+        logger.info("block");
+        setBlocking(true);
+
+        // wait
+        logger.info("in-flight");
+        Object response = getDelegate().cmd(MedusaBroadcast2NodesDAO.IN_FLIGHT_CMD);
+        if ( response != null ) {
+          Long inFlight = (Long) response;
+          if ( inFlight > 0L ) {
+            // TODO: wait
+            logger.warning("in-flight", inFlight);
+          }
+        }
+
+        // nodes
+        logger.info("nodes");
+        rollNodes(x);
+
+        // dagger
+        logger.info("dagger");
+        long oldIndex = dagger(x);
+
+        // unblock
+        logger.info("unblock");
+        setBlocking(false);
+        super.unblock(x);
+
+        // compaction
+        logger.info("compaction");
+        compaction(x);
+
+        // purge
+        logger.info("purge");
+        purge(x, oldIndex);
+
+        logger.info("end");
+      } catch (Throwable t) {
+        logger.error(t);
+      }
+      `
+    },
+    {
+      name: 'rollNodes',
+      args: 'X x',
+      javaCode: `
+      final Logger logger = Loggers.logger(x, this);
+      // logger.info("rollNodes");
+
+      final ClusterConfigSupport support = (ClusterConfigSupport) x.get("clusterConfigSupport");
+      final ClusterConfig myConfig = support.getConfig(x, support.getConfigId());
+
+      AssemblyLine line = new AsyncAssemblyLine(x, null, support.getThreadPoolName());
+      final Set replies = new HashSet();
+      List<ClusterConfig> nodes = support.getReplayNodes();
+      for ( ClusterConfig cfg : nodes ) {
+        line.enqueue(new AbstractAssembly() {
+          public void executeJob() {
+            DAO client = support.getClientDAO(x, "medusaEntryDAO", myConfig, cfg);
+            try {
+              replies.add(client.cmd(new FileRollCmd()));
+            } catch (RuntimeException e) {
+              logger.error("rollNodes", "cmd", e);
+            }
+          }
+        });
+      }
+
+      logger.debug("rollNodes", "line.shutdown", "wait");
+      // NOTE: this will block forever if a node does not reply.
+      // line.shutdown();
+      // Perform manual polling for completion.
+      long waited = 0L;
+      long sleep = 1000L;
+      while ( waited < getMaxWait() ) {
+        try {
+          Thread.currentThread().sleep(sleep);
+          waited += sleep;
+        } catch (InterruptedException e) {
+          break;
+        }
+        if ( replies.size() >= nodes.size() -1 ) {
```

Comment:
why here is `>= nodes.size() -1 ` not `>= nodes.size()`

Reply:
corrected.